src/libaducpal/src/ftw.c

```diff
@@ -165,318 +78,124 @@ static int construct_typeflag(const DWORD attrib, int flags)
     return typeflag;
 };
 
-/* this function was adapted from a public domain Rosetta Code example:
- * https://rosettacode.org/wiki/Walk_a_directory/Recursively#Windows
- */
-static int do_wnftw(wchar_t* wpath, NFTW_FUNC_T func, int fd_limit, int flags, struct history* h)
+static int do_callback(NFTW_FUNC_T func, const char* fpath, int flags, struct FTW* ftwbuf)
 {
-    struct stack
-    {
-        wchar_t* path;
-        size_t pathlen;
-        size_t slashlen;
-        HANDLE ffh;
-        WIN32_FIND_DATAW ffd;
-        struct stack* next;
-    } *dir, dir0, *ndir;
-    size_t patternlen;
-    wchar_t *buf, c, *pattern;
-    int rval = 0;
-
-    dir0.path = wpath;
-    dir0.pathlen = wcslen(dir0.path);
-    pattern = L"*";
-    patternlen = wcslen(pattern);
-
-    //
-    // Must put backslash between path and pattern, unless
-    // last character of path is slash or colon.
-    //
-    //   'dir' => 'dir\*'
-    //   'dir\' => 'dir\*'
-    //   'dir/' => 'dir/*'
-    //   'c:' => 'c:*'
-    //
-    // 'c:*' and 'c:\*' are different files!
-    //
-    c = dir0.path[dir0.pathlen - 1];
-    if (c == ':' || c == '/' || c == '\\')
+    struct _stat st64i32;
+    if (_stat(fpath, &st64i32) != 0)
     {
-        dir0.slashlen = dir0.pathlen;
-    }
-    else
-    {
-        dir0.slashlen = dir0.pathlen + 1;
+        return -1;
     }
 
-    /* Allocate space for path + backslash + pattern + \0. */
-    buf = calloc(dir0.slashlen + patternlen + 1, sizeof buf[0]);
-    if (buf == NULL)
+    struct stat st;
+    wstat_to_stat(&st, &st64i32);
+
+    const int typeflag = typeflag_from_stat(&st, flags);
+
+    // If func() returns nonzero, then the tree walk is terminated and the value returned by fn()
+    // is returned as the result of nftw().
+    return func(fpath, &st, typeflag, ftwbuf);
+}
+
+// Return 0 on success, -1 if an error occurs.
+static int do_nftw(const char* path, NFTW_FUNC_T func, int nopenfd, int flags, int level)
```

Comment:
I would expect the impl to start closing older FDs when it gets past some threshold so that it doesn't keep too many FDs open.

Reply:
I may have missed it, but I don't see any logic ensuring nopenfd as a max number of FDs to have opened--therefore, the process could run out of FDs.  It's supposed to start closing older FDs that age out of the FD buffer with `nopenfd` number of elements as it recurses down the directory tree.

As per the manpage https://linux.die.net/man/3/nftw:
> To avoid using up all of the calling process's file descriptors, nopenfd specifies the maximum number of directories that ftw() will hold open simultaneously. When the search depth exceeds this, ftw() will become slower because directories have to be closed and reopened. ftw() uses at most one file descriptor for each level in the directory tree.

Reply:
yeah, i'll add a TODO for respecting nopenfd - I meant to put that in there.  I don't think it's critical for us right now.